plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTaskService.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2021 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.gradle.spotless;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.api.tasks.Internal;
+
+import com.diffplug.common.base.Preconditions;
+import com.diffplug.common.base.Unhandled;
+
+/**
+ * Allows the check and apply tasks to coordinate
+ * with each other (and the source task) to reduce
+ * duplicated work (e.g. no need for check to run if
+ * apply already did).
+ */
+public abstract class SpotlessTaskService implements BuildService<BuildServiceParameters.None> {
+	private final Map<String, SpotlessApply> apply = Collections.synchronizedMap(new HashMap<>());
+	private final Map<String, SpotlessTask> source = Collections.synchronizedMap(new HashMap<>());
```

Comment:
Have you considered using `ConcurrentMap` over synchronized maps here? I ask because of we ever need to loop, stream, or otherwise do some iteration operation over either of these maps (such an operation may not be obvious), then the map would need to be `synchronized` first to prevent concurrency problems.

Reply:
I think it's gonna stay pretty simple.

Reply:
Let me clarify -> tasks don't need to coordinate *across* projects, they only coordinate *within* a single project. If we could specify a `SpotlessTaskService` per-project, then we wouldn't need it to be synchronized. So concurrent access is possible, but I see no risk of deadlock nor of inconsistent state, because the parts that need to coordinate are single-threaded,.

Reply:
Ah, gotcha. I'd still argue that `ConcurrentMap` is the better solution, but it looks like we'll have to agree to disagree. So I'm happy to defer to you once more. :)